### ISSUE TITLE
Shared subscriptions don't appear to work as intended

### ISSUE THREAD

**Issue by oakad** [erlio/vernemq#370]
Experienced with vernemq 1.0.0 and 1.0.1

Assume client cl0 subscribes to "$share/xx/t0" and some other client publishes to topic "t0". Nothing happens (client cl0 gets no messages).

Assume another clients assumes id cl0 and subscribes to topic "t0". Now both clients (the one with subscription to "$share/xx/t0" and the one with subscription to "t0") get messages in load-balanced fashion (similar to old, non-standard behavior).

Desired behavior: multiple clients with unique ids subscribed to "$share/xx/t0" participate in load balanced deliveries of messages published to "t0".

**Comment by larshesel** [comment]
Hi - I can't reproduce this - could you provide a bit more information? How are you connecting to vernemq (load balancers)? Are you running on multiple nodes? What is your shared subscription policy?

Can you reproduce this on a single node? Or only when running in a cluster? Do you see any errors in the logs?

**Comment by oakad** [comment]
No errors, no clustering, single node only. A log is mostly filled with messages like the one below and nothing else:

2017-04-21 01:16:26.789 [debug] <0.221.0>@plumtree_broadcast:schedule_lazy_tick:
602 0ms mailbox traversal, schedule next lazy broadcast in 10000ms, the min inte
rval is 10000ms


A part of config file goes like that:

allow_anonymous = off
allow_register_during_netsplit = off
allow_publish_during_netsplit = off
allow_subscribe_during_netsplit = off
allow_unsubscribe_during_netsplit = off
allow_multiple_sessions = on
retry_interval = 20
max_client_id_size = 100
persistent_client_expiration = 1w
max_inflight_messages = 20
max_online_messages = 1000
max_offline_messages = 1000
message_size_limit = 0
upgrade_outgoing_qos = off
queue_deliver_mode = balance
shared_subscription_policy = prefer_local

My testing:
Attempt 1:
mach0> mosquitto_sub -u mqtt_tester -P mqtt_tester -h 10.112.0.66 -p 1883 -t $share/xx/t1 -i cl.001
mach1> mosquitto_pub -u mqtt_tester -P mqtt_tester -h 10.112.0.66 -p 1883 -t t1 -i cl.002 -m aaaa

Message not delivered. From now on, assume that mach1 just keeps publishing in a loop.

Attempt 2:
mach2> mosquitto_sub -u mqtt_tester -P mqtt_tester -h 10.112.0.66 -p 1883 -t t1 -i cl.001

Suddenly messages are delivered to either mach0 and mach2, with mach2 getting somewhat larger amount of messages. Mach2 uses same id as mach0.

Attempt 3:
Let's kill the mach2 subscriber. All messages go to mach0. This is correct now, but was not working in Attempt 1.

Attempt 4:
mach2> mosquitto_sub -u mqtt_tester -P mqtt_tester -h 10.112.0.66 -p 1883 -t $share/xx/t1 -i cl.003

Mach2 has got an unique id, yet mach2 get's no messages. All messages go to mach0.

Attempt 5:

mach2> mosquitto_sub -u mqtt_tester -P mqtt_tester -h 10.112.0.66 -p 1883 -t $share/xx/t1 -i cl.001

Give mach2 same id as mach0. Either mach2 or mach0 get the message (kind of correct, but both subscribers are stuck with the same id).

Attempt 6:
Kill both mach0 and mach2 publishers. Go back to Attempt 1. The wrong behavior stays.

Expected behavior:
1. Attempt 1 works right every time
2. Members of the shared subscription should be able to have unique ids
3. Attempt 2 looks like buggy behavior outright (older load balancing method somehow making the shared subscriptions work)

**Comment by oakad** [comment]
It appears now that our netty java client may be to blame for the issues we are seeing (and my previous post may suffer from improper shell quoting of "$share" identifier). I will therefore beg your pardon and close this report for a time being in order to get better understanding of the client side issues we may be facing.

**Comment by larshesel** [comment]
I think the `$` in the topic of a shared subscription is quite annoying and  it's really easy to run into shell issues. I'll add a note about this pitfall to our documentation.

And no worries! Please let us know if you encounter other issues!

**Comment by oakad** [comment]
After some careful investigation, we, nevertheless, came to a conclusion that there exists some problem with shared subscriptions in vmq. Namely, we observed the following 2 patterns:

1. $share/... subscribers don't get any messages at all
2. $share/.../<topic> subscribers behave as if they were normal subscribers on the <topic> - each subscriber gets a copy of message

This behaviors appeared to result from arbitrary subscribe/unsubscribe fiddling by peers.

However, when we deleted the vmq state (under "/var/lib/vernemq") the strange behavior was gone and the vmq was operating up to spec.

This is rather worrisome, because it points at the possibility, that vmq state may become corrupted in production.

Another thing: our original install was 1.0.0, which was upgraded to 1.0.1. May this be a cause for corruption? If yes, vmq must complain about leveldb version mismatch.

**Comment by ioolkos** [comment]
Hi @oakad thanks for reporting!

Is it possible to give any more information on the "arbitrary subscribe/unsubscribe fiddling by peers".
Did the clients change anything, during subscribing (QoS, clean session false)? You're still using mosquitto_sub here?

**Comment by oakad** [comment]
Actually, we are using a custom java client using netty's mqtt codec support (https://netty.io/4.1/api/io/netty/handler/codec/mqtt/package-summary.html); we used mosquitto_sub/pub for manual testing to confirm the issues. That java client had some problems of its own, but the individual messages it sent were well-formed. We limited ourselves to:

1. QoS level 0
2. Always clean sessions, no last testament
3. No message retention

At times, we made multiple clients to connect with the same id's, but mostly they all had unique ids (sometimes longer that 23 chars, sometimes shorter; now we use only longer ids, because they are derived from AWS' EC2 instance ids, which are somewhat long these days). All clients used the same username/password pair (using same vmq.passwd entry). "Shared" subscribers subscribed to wildcard topic with a single "+" in it (either in the middle or at the end).

We also observed some rare message corruptions in the logs (while vmq was still misbehaving).

After wiping the db directory clean and restarting vmq all this behavior was gone instantly and we have not managed to trigger it again on Friday (but we are still testing).

**Comment by oakad** [comment]
Regarding the corruption issue - that's the log message we are seeing (the mqtt message is probably lost afterwards):

<28>0 2017-05-01T04:18:57.312615+00:00 ip-10-112-0-136 vmq 9321 - - [warning] <0.6383.10>@vmq_ranch:teardown:127 session stopped abnormally due to '{error,unexpected_message,{error,cant_parse_variable_header}}'

This happens only rarely and under higher load. We have no idea what message is this and we are fairly certain that the clients are not playing tricks with the headers. Is there a way to force vmq to dump the offending message into the log?

**Comment by ioolkos** [comment]
Hi @oakad thanks a lot for your efforts! 
As a quick confirmation: you're using protocol version 3.1, not 3.1.1? at least that's what the netty codec seems to indicate. Verne should be compatible to 3.1, but maybe there's a subtle difference at play here.

**Comment by oakad** [comment]
We are using 3.1.1 of course - netty codec supports both.

**Comment by larshesel** [comment]
Regarding the malformed/unparseable there currently aren't any way to dump the bad message. It would be good to inspect he data using tcpdump/wireshark to see if something looks corrupted in the data flow.

On which platform are you experiencing these issues?

**Comment by oakad** [comment]
AWS/Amazon Linux
The wireshark approach is not very constructive as those issues happen at substantial load (thousands of messages per second).

**Comment by ioolkos** [comment]
OK, I'm currently trying to collect all the information you've given so far in my "case file" for cross referencing and see if it gives me more ideas.

More info would still be helpful of course (what message rates are you trying to reach? what Amazon instance type?)
It's also unclear to me, if you actually can reproduce both faulty behaviours here https://github.com/erlio/vernemq/issues/370#issuecomment-298801995 with a low number of mosquitto_pub/sub commands, or not.

Again, thanks for your efforts.

**Comment by oakad** [comment]
That behavior was gone after we deleted the /var/lib/vernemq dir and restarted the service. We have to assume it was caused by db corruption of some sort.

The instance type is c3.8xlarge, 32 CPU machine. We want to exceed 50k simultaneous subscribers per machine and got to 20k publishes/sec.

Basically, our project is to try to replace a current HTTP based service, serving up to 5k rps with semi-persistent mqtt connections (the whole thing is used to support our fairly popular games).
